app/uk/gov/hmrc/gform/fileupload/FileUploadController.scala

```diff
@@ -106,18 +106,35 @@ class FileUploadController(
 
   private def validateFile(fileId: FileId, envelope: Envelope)(implicit messages: Messages): Validated[Flash, Unit] =
     envelope.find(fileId).fold[Validated[Flash, Unit]](Valid(())) { file =>
+      val fileExtension = getFileExtension(file.fileName)
+      val fileExtensionCheck = fileExtension.fold(false) { v =>
+        !appConfig.restrictedFileExtensions.map(_.value).contains(v.toUpperCase)
+      }
+      val fileTypeCheck = appConfig.contentTypes.exists(_ === file.contentType)
+
       Valid(file)
         .ensure(mkFlash("file.error.empty"))(_.length =!= 0)
         .ensure(
-          mkFlash("file.error.type", file.contentType.value, "ODS, XLSX, DOCX, ODT, PPTX, ODP, PDF, JPEG")
-        )(file => appConfig.contentTypes.exists(_ === file.contentType))
+          mkFlash(
+            "file.error.type",
+            if (!fileExtensionCheck) fileExtension.getOrElse("UNKNOWN") else file.contentType.value,
+            "PDF, JPEG, XLSX, ODS, DOCX, ODT, PPTX, ODP"
+          )
+        )(_ => fileExtensionCheck && fileTypeCheck)
```

Comment:
would have been cleaner to defined fileExtensionCheck and fileTypeCheck as local methods that accept file param?